### DISCUSSION TITLE
Brew restoring old `python3.9` link after upgrade

### DISCUSSION THREAD

**Discussion by SRv6d** [Homebrew/discussions#3363] [Everyday usage]
I have `python3.9` and `python3.10` installed. `python3.10` I need for my daily work and `python3.9` exists due to dependencies:

> Error: Refusing to uninstall /opt/homebrew/Cellar/python@3.9/3.9.13_1
because it is required by cairo, ffmpeg, glib, gobject-introspection, harfbuzz, ldns, libass and openssh, which are currently installed.

I've linked `python3.10` to be the default version using `brew link --overwrite python@3.10`, but more often than not, after running brew upgrade, brew restores the old link to `python3.9`. Is it possible to make this link persist after an upgrade ?

**Comment by superatomic** [comment]
I have been encountering the same problem for quite some time and can report the same exact behavior. `brew link --overwrite python@3.10` will make correctly link `python`, but it will eventually be reset.

I believe this is caused by the formula `python` being an alias to `python@3.9` instead of `python@3.10`.

Here are the relevant symlinks in `homebrew/core`:
- [`python`](https://github.com/Homebrew/homebrew-core/blob/master/Aliases/python)
- [`python3`](https://github.com/Homebrew/homebrew-core/blob/master/Aliases/python3)
- [`python@3`](https://github.com/Homebrew/homebrew-core/blob/master/Aliases/python%403)

The relevant pull request that may resolve this problem is here:
- https://github.com/Homebrew/homebrew-core/pull/87075

The PR has been open for a while now, but it is still active.

**Reply by superatomic** [reply]
As a workaround, try adding `/usr/local/opt/python@3.10/bin` to your PATH, making sure to place it before Homebrew's own `bin` directory. On non-Intel Macs this path might be different.
